File: templates/trie/trie.go
package trie

type Node struct {
	children [26]*Node
	isWord   bool
}

type Trie struct {
	root *Node
}

func NewTrie() *Trie {
	return &Trie{root: &Node{}}
}

func (t *Trie) Insert(word string) {
	cur := t.root
	for i := 0; i < len(word); i++ {
		idx := word[i] - 'a'
		if cur.children[idx] == nil {
			cur.children[idx] = &Node{}
		}
		cur = cur.children[idx]
	}
	cur.isWord = true
}

func (t *Trie) Search(word string) bool {
	node := t.walk(word)
	return node != nil && node.isWord
}

func (t *Trie) StartsWith(prefix string) bool {
	return t.walk(prefix) != nil
}

func (t *Trie) walk(s string) *Node {
	cur := t.root
	for i := 0; i < len(s); i++ {
		idx := s[i] - 'a'
		if cur.children[idx] == nil {
			return nil
		}
		cur = cur.children[idx]
	}
	return cur
}

File: templates/trie/trie.py
class Node:
    def __init__(self) -> None:
        self.children: dict[str, Node] = {}
        self.is_word = False


class Trie:
    def __init__(self) -> None:
        self.root = Node()

    def insert(self, word: str) -> None:
        current = self.root
        for ch in word:
            current = current.children.setdefault(ch, Node())
        current.is_word = True

    def search(self, word: str) -> bool:
        node = self._walk(word)
        return node is not None and node.is_word

    def starts_with(self, prefix: str) -> bool:
        return self._walk(prefix) is not None

    def _walk(self, text: str) -> Node | None:
        current = self.root
        for ch in text:
            if ch not in current.children:
                return None
            current = current.children[ch]
        return current
